File: test/itemSurfaceFlagsSeed.mjs
const PLACEABLE_ITEM_TYPES = new Set([
  'flat-panel',
  'base',
  'counter',
  'separator',
  'sofa-set-classic',
  'sofa-single-classic',
  'sofa-double-classic',
  'coffee-table-classic',
  'table-chair-set-eames',
  'chair',
  'table-glass',
  'bar-stool',
  'mini-fridge',
  'kettle',
  'coat-rack',
  'upright',
  'profile',
  'plastic-trash-bin',
  'box-block',
  'indoor-plant-1',
  'tv',
  'shelf',
  'led-floodlight',
  'door',
  'showcase-2',
  'showcase-3',
  'illuminated-foam',
]);

const IMAGE_COLOR_TYPES = new Set([
  'door-leaf',
  'flat-panel',
  'base',
  'counter',
  'door',
  'showcase-2',
  'showcase-3',
]);

const PANEL_COVER_TYPES = new Set([
  'flat-panel',
  'showcase-2',
  'showcase-3',
  'door',
]);

const COLOR_ONLY_TYPES = new Set([
  'separator',
  'sofa-set-classic',
  'sofa-single-classic',
  'sofa-double-classic',
]);

const LEAF_COLOR_IMAGE_TYPES = new Set(['door-leaf']);

/** Parent içinde çizilen parçalar. Kendi modül fabrikası yok; katalog şart değil. */
const EMBEDDED_RENDER_TYPES = new Set([
  'panel',
  'separator-panel',
  'base-top',
  'counter-top',
  'connector',
  'showcase-board',
  'showcase-accessory',
  'video-wall-panel',
]);

const EMBEDDED_COLOR_TYPES = new Set([
  'panel',
  'separator-panel',
  'base-top',
  'counter-top',
]);

const EMBEDDED_COVER_TYPES = new Set(['panel']);

const ALL_FALSE = Object.freeze({
  isRender: false,
  acceptsColor: false,
  acceptsImage: false,
  acceptsLightbox: false,
  acceptsGlass: false,
  acceptsMesh: false,
});

export function surfaceFlagsForItem(itemKey, itemType) {
  void itemKey;
  if (LEAF_COLOR_IMAGE_TYPES.has(itemType) || EMBEDDED_RENDER_TYPES.has(itemType)) {
    return {
      isRender: true,
      acceptsColor: LEAF_COLOR_IMAGE_TYPES.has(itemType) || EMBEDDED_COLOR_TYPES.has(itemType),
      acceptsImage: LEAF_COLOR_IMAGE_TYPES.has(itemType) || EMBEDDED_COVER_TYPES.has(itemType),
      acceptsLightbox: EMBEDDED_COVER_TYPES.has(itemType),
      acceptsGlass: EMBEDDED_COVER_TYPES.has(itemType),
      acceptsMesh: EMBEDDED_COVER_TYPES.has(itemType),
    };
  }
  if (!PLACEABLE_ITEM_TYPES.has(itemType)) return { ...ALL_FALSE };
  const cover = PANEL_COVER_TYPES.has(itemType);
  const imageColor = IMAGE_COLOR_TYPES.has(itemType);
  const colorOnly = COLOR_ONLY_TYPES.has(itemType);
  return {
    isRender: true,
    acceptsColor: imageColor || colorOnly,
    acceptsImage: imageColor,
    acceptsLightbox: cover,
    acceptsGlass: cover,
    acceptsMesh: cover,
  };
}

export function applyItemSurfaceFlags(item) {
  Object.assign(item, surfaceFlagsForItem(item.itemKey, item.type));
  return item;
}

File: backend/app/modules/fair_stand/infrastructure/item_surface_flags_seed.py
"""Item is_render + accepts_* seed. Matches current scene gates, not future wiring."""

from __future__ import annotations

import sqlalchemy as sa

from app.modules.fair_stand.infrastructure.item_rotation_seed import PLACEABLE_ITEM_TYPES

# itemCapabilities.js COLOR_AND_IMAGE types (image+color today).
_IMAGE_COLOR_TYPES = frozenset(
    {
        "door-leaf",
        "flat-panel",
        "base",
        "counter",
        "door",
        "showcase-2",
        "showcase-3",
    }
)

# pickModuleContext: selectionMode === "panel" (glass/lightbox/mesh menu).
_PANEL_COVER_TYPES = frozenset(
    {
        "flat-panel",
        "showcase-2",
        "showcase-3",
        "door",
    }
)

_COLOR_ONLY_TYPES = frozenset(
    {
        "separator",
        "sofa-set-classic",
        "sofa-single-classic",
        "sofa-double-classic",
    }
)

_LEAF_COLOR_IMAGE_TYPES = frozenset({"door-leaf"})


def surface_flags_for_item(item_key: str, item_type: str) -> dict[str, bool]:
    del item_key
    if item_type in _LEAF_COLOR_IMAGE_TYPES:
        return {
            "is_render": True,
            "accepts_color": True,
            "accepts_image": True,
            "accepts_lightbox": False,
            "accepts_glass": False,
            "accepts_mesh": False,
        }
    is_render = item_type in PLACEABLE_ITEM_TYPES
    if not is_render:
        return {
            "is_render": False,
            "accepts_color": False,
            "accepts_image": False,
            "accepts_lightbox": False,
            "accepts_glass": False,
            "accepts_mesh": False,
        }

    cover = item_type in _PANEL_COVER_TYPES
    image_color = item_type in _IMAGE_COLOR_TYPES
    color_only = item_type in _COLOR_ONLY_TYPES
    return {
        "is_render": True,
        "accepts_color": image_color or color_only,
        "accepts_image": image_color,
        "accepts_lightbox": cover,
        "accepts_glass": cover,
        "accepts_mesh": cover,
    }


def apply_item_surface_flags(row: dict) -> dict:
    flags = surface_flags_for_item(row["item_key"], row["item_type"])
    row.update(flags)
    return row


def fill_item_surface_flag_columns(bind) -> None:
    existing = bind.execute(sa.text("SELECT item_key, item_type FROM fair_stand_items")).mappings()
    for row in existing:
        flags = surface_flags_for_item(row["item_key"], row["item_type"])
        bind.execute(
            sa.text(
                "UPDATE fair_stand_items SET is_render = :is_render, "
                "accepts_color = :accepts_color, accepts_image = :accepts_image, "
                "accepts_lightbox = :accepts_lightbox, accepts_glass = :accepts_glass, "
                "accepts_mesh = :accepts_mesh WHERE item_key = :item_key"
            ),
            {**flags, "item_key": row["item_key"]},
        )
